brainiak/funcalign/fastsrm.py

```diff
@@ -0,0 +1,1014 @@
+
+    n_components : int
+        Number of timecourses of the shared coordinates
+
+    n_iter : int
+        Number of iterations to perform
+
+    temp_dir : str or None
+        path to dir where temporary results are stored
+        if None temporary results will be stored in memory. This
+        can results in memory errors when the number of subjects
+        and / or sessions is large
+
+    low_ram : bool
+        if True and temp_dir is not None, reduced_data will be saved on disk
+        this increases the number of IO but reduces memory complexity when
+        the number of subject and / or sessions is large
+
+    random_state : int or RandomState
+        Pseudo number generator state used for random sampling.
+
+    n_jobs : int, optional, default=1
+        The number of CPUs to use to do the computation.
+         -1 means all CPUs, -2 all CPUs but one, and so on.
+
+    verbose : bool or "warn"
+        if True, logs are enabled.
+        if False, logs are disabled.
+        if "warn" only warnings are printed.
+
+    Attributes
+    ----------
+
+    `basis_list`: list of array, element i has shape=[n_components, n_voxels]
+     or list of str
+        basis of all subjects, element i is the basis of subject i
+        or path to basis of all subjects, element i is the path to the
+        basis of subject i
+    """
+    def __init__(self,
+                 atlas,
+                 n_components=20,
+                 n_iter=100,
+                 temp_dir=None,
+                 low_ram=False,
+                 random_state=None,
+                 n_jobs=1,
+                 verbose="warn",):
+
+        self.random_state = random_state
+        self.n_jobs = n_jobs
+        self.verbose = verbose
+        self.n_components = n_components
+        self.n_iter = n_iter
+        self.atlas = atlas
+
+        self.basis_list = None
+
+        if temp_dir is None:
+            if self.verbose == "warn" or self.verbose is True:
+                logger.warning("temp_dir has value None. "
+                               "All basis (spatial maps) and reconstructed "
+                               "data will therefore be kept in memory."
+                               "This can lead to memory errors when the "
+                               "number of subjects "
+                               "and/or sessions is large.")
+            self.temp_dir = None
+            self.low_ram = False
+
+        if temp_dir is not None:
+            if not os.path.exists(os.path.join(temp_dir, "fastsrm")):
+                os.mkdir(os.path.join(temp_dir, "fastsrm"))
+            self.temp_dir = os.path.join(temp_dir, "fastsrm")
+
+            # Remove files in temp folder
+            paths = glob.glob(os.path.join(self.temp_dir, "*.npy"))
+            for path in paths:
+                os.remove(path)
+
+            self.low_ram = low_ram
+
+    def fit(self, imgs):
+        """Computes basis across subjects from input imgs
```

Comment:
Optional: You could consider providing an API option that is the same as SRM, where the input is a list of arrays of size voxel-by-time. Also, the input data of FastSRM has a different orientation compared with all the other algorithms in the funcalign module. It might be better to keep everything consistent.

Reply:
I'll add this option. However being able to provide a list of .npy file instead of a list of arrays allows FastSRM to handle datasets that are difficult to hold in memory so I would like to keep this possibility.

Reply:
Yes, keeping both options would be optimal.